### PR TITLE
Add field `default` to Parameter

### PR DESCRIPTION
**Pull request by nstarman**
Split off from #15168. This adds the field `default` to Parameter. Currently it doesn't do anything since the default value is specified on `__init__()`, but it is useful for introspecting a cosmology class e.g. `LambdaCDM.H0.default` and it will also replace `__init__()` when `Cosmology` is a dataclass for v6.

### CODE REVIEW

**Comment by nstarman** [comment]
Let's move the discussion about `NotImplemented` to here.

**Comment by nstarman** [comment]
And for comparison... I'm adding a 2nd commit with the sentinel value

File: docs/changes/cosmology/15400.feature.rst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Cosmology I/O now auto-identifies the '.tex' suffix with the 'ascii.latex' format. The
```
Comment (nstarman) [inline docs/changes/cosmology/15400.feature.rst:1]: Oops. I'll remove this before we merge.

```suggestion
The
```

File: astropy/cosmology/parameter/_core.py [hunk 2]
```diff
@@ -129,12 +149,14 @@ class Parameter:
         def __init__(
             self,
             *,
+            default: Any | Literal[Sentinel.has_no_default] = HASNODEFAULT,
```
Comment (nstarman) [inline astropy/cosmology/parameter/_core.py:152]: I'll remove the type hint before we merge.

```suggestion
            default = HASNODEFAULT,
```

**Comment by nstarman** [comment]
Just to say both passed.

CleanShot 2023-09-27 at 18 39 51@2x

**Comment by eerovaher** [comment]
Can you explain in more detail why `HASNODEFAULT` is needed and why wouldn't it be possible to use `None` as the sentinel value instead?

**Comment by nstarman** [comment]
If `None` is the sentinel value then it can't be used as a real value, which it is for `Ob0`.

**Comment by nstarman** [comment]
`inspect`, for example, uses `_missing` as their custom sentinel value for the same reason.

**Comment by nstarman** [comment]
[Python needs Sentinels](https://peps.python.org/pep-0661/)

**Comment by nstarman** [comment]
In commit #1 I had `NotImplemented` as the sentinel.

**Comment by mhvk** [comment]
One could also consider `np._NoValue` for the sentinel.

**Comment by nstarman** [comment]
> One could also consider `np._NoValue` for the sentinel.

I think we should avoid private api. Otherwise `insect._empty`, used by `inspect.Signature` for this purpose, would be best.

**Comment by mhvk** [comment]
The advantage of `np._NoValue` over `inspect._empty` is that users are not unlikely to have seen it in documentation, and that it overrides `__repr__` to be "<no value>". But, yes, nominally private API. I think I'd still pick it over a homebrew one, but if you go with yours I would suggest to add a `__repr__` (maybe "<no default>"?) so that the sphinx docs look OK.

p.s. PEP 661 was interesting to read, thanks for the link!

**Comment by nstarman** [comment]
> The advantage of `np._NoValue` over `inspect._empty` is that users are not unlikely to have seen it in documentation, and that it overrides `__repr__` to be "".
> But, yes, nominally private API. I think I'd still pick it over a homebrew one, but if you go with yours I would suggest to add a `__repr__` (maybe ""?) so that the sphinx docs look OK.

I like the the `repr`! But I the values in Cosmology instances don't have to by NumPy objects, so I don't want to use a numpy object as the sentinel placeholder.  But I will  definitely do the repr!

I think I'll also name the object ``MISSING``, which is in line with https://docs.python.org/3/library/dataclasses.html#dataclasses.MISSING and `inspect._empty`

> p.s. PEP 661 was interesting to read, thanks for the link!

👍

**Comment by nstarman** [comment]
I changed the repr to ``"<MISSING>"`` which is better than what `Enum` was doing. Thanks @mhvk !

**Comment by eerovaher** [comment]
> If `None` is the sentinel value then it can't be used as a real value, which it is for `Ob0`.

`Ob0` is the ratio of baryonic matter density and critical density at `z=0`. I'd expect that value to be a `float` and I don't understand how `None` could be a valid ratio. If the default value of `Ob0` would be a `float`, would `MISSING` still be needed?

**Comment by eerovaher** [comment]
What if the default for `Ob0` would be `math.nan` or `np.nan`?

**Comment by nstarman** [comment]
Interesting! I hadn't considered that. My hesitations are 2-fold:
1. I don't want users to have to differentiate between math and numpy's NaN.

CleanShot 2023-10-03 at 12 36 41@2x

2. This doesn't solve the underlying issue of multi-modality. I would much rather have a separate class that has matter only (without baryons & DM). So then `None` wouldn't be needed for `Ob0`, but for a different reason. 

3. We let users define custom cosmologies. `None` is sufficiently common that I don't really want to disallow it as a default value for a Parameter. It seems unnecessary to force anyone implementing a similar pattern to our `Ob0` to also have to change. Just because I don't like the multimodality doesn't mean it isn't a decent solution in another context. It's for similar reasons that Python made `MISSING` in `dataclasses` and `_empty` in `inspect`. It all comes back to the fact that Python forces us to roll our own Sentinels, as is done in this PR.

**Comment by nstarman** [comment]
@eerovaher, do you agree with this reasoning? This code pattern is suggested in https://peps.python.org/pep-0484/#support-for-singleton-types-in-unions for how to build statically recognizable sentinel values (until the PEP for Sentinels is in for py3.1X).

**Comment by eerovaher** [comment]
I don't use `cosmology`, so if there are any quirks because of historical reasons then I am blissfully unaware of them. As far as I can tell the appropriate data type for all the parameter values is `float` (or `ndarray` of floats or `Quantity` or something similar). If a sentinel value is needed then it should be possible to use `nan`, which is a valid `float`, or `None`, which is a Python builtin commonly used for this purpose. It is not true that `math.nan` is difficult to use with `numpy` or vice versa:
```python
>>> import math
>>> import numpy as np
>>> math.isnan(np.nan)
True
>>> np.isnan(math.nan)
True
```
So in short I still don't understand why a brand new sentinel value is needed here. The reason `dataclasses` or `inspect` have their own sentinel value is that they don't have the luxury of knowing what data types they will be working with. I wouldn't be recommending `nan` as a sentinel for strings.

**Comment by nstarman** [comment]
> It is not true that math.nan is difficult to use with numpy or vice versa:

They are statically different and also floats cannot be differentiated statically. I think that's important.

```python
@overload
def function(x: Literal[inf]) -> NoReturn:  # Not a valid entry to `Literal`
     ...

@overload
def function(x: float) -> np.ndarray:
     ...

def function(x: float) -> np.ndarray:
     if x is inf: 
        raise ValueError
     return ...
```

 So I actually like `None` as a sentinel better than `nan`. But `None` can't be used because of cosmology's historical quirks (a good turn of phrase). I think this actually ends up being a minor point since users will never see this unless they are deep-diving into cosmology's internals. If / when cosmology's quirks are cleaned up then `MISSING` can be deprecated in favor of `None` (a suitably long deprecation period in which both are valid)

File: astropy/cosmology/parameter/_core.py [hunk 3]
```diff
@@ -275,8 +308,13 @@ def clone(self, **kw):
...
         fields_repr = (
+            # Get the repr, using the input fvalidate over the processed value
             f"{f.name}={(getattr(self, f.name if f.name != 'fvalidate' else '_fvalidate_in'))!r}"
             for f in fields(self)
-            if f.repr
+            # Only show fields that should be displayed and are not sentinel values
+            if (
+                f.repr
+                and (self.default is not MISSING if f.name == "default" else True)
+            )
```
Comment (eerovaher) [inline astropy/cosmology/parameter/_core.py:318]:
```suggestion
            if (
                f.repr 
                and (f.name != "default" or self.default is not MISSING)
            )
```

File: astropy/cosmology/tests/test_parameter.py [hunk 4]
```diff
@@ -37,6 +37,42 @@ def test_registry_validators():
...
+            doc="DOCSTRING",
+            unit="km",
+            equivalencies=[u.mass_energy()],
+            derived=True,
+        )
+        assert parameter.default == 1.0
+        assert parameter.fvalidate is _validate_to_float
+        assert parameter.unit is u.km
+        assert parameter.equivalencies == [u.mass_energy()]
+        assert parameter.derived is True
```
Comment (eerovaher) [inline astropy/cosmology/tests/test_parameter.py:67]: Wouldn't this be better as a parametrized test?

File: astropy/cosmology/tests/test_parameter.py [hunk 5]
```diff
@@ -139,15 +152,21 @@ def test_Parameter_derived(self, cosmo_cls, all_parameter):
...
     # -------------------------------------------
     # descriptor methods
 
     def test_Parameter_descriptor_get(self, cosmo_cls, cosmo, all_parameter):
         """Test :attr:`astropy.cosmology.Parameter.__get__`."""
         # from class
         parameter = getattr(cosmo_cls, all_parameter.name)
-        assert isinstance(parameter, Parameter)
-        assert parameter is all_parameter
+        assert np.all(parameter.default == all_parameter.default)
```
Comment (eerovaher) [inline astropy/cosmology/tests/test_parameter.py:169]: Test shouldn't be using `assert np.all(...)`. They should be using either [`np.testing.assert_array_equal()`](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_array_equal.html), which produces better output if the test fails, or `assert all(...)` if `assert_array_equal()` can't be used.

File: docs/whatsnew/6.0.rst [hunk 6]
```diff
@@ -140,7 +140,7 @@ This means that the :mod:`dataclasses` machinery can be used to work with
 
     >>> from dataclasses import asdict
     >>> asdict(H0)
-    {'derived': False, 'unit': Unit("km / (Mpc s)"), 'equivalencies': [], ...
+    {'default': ..., 'derived': False, 'unit': Unit("km / (Mpc s)"), 'equivalencies': [], ...
```
Comment (eerovaher) [inline docs/whatsnew/6.0.rst:143]: Why is the `'default'` value not written out?

**Review by eerovaher — COMMENTED** [review]
I understand now why a brand new sentinel value is needed. It would be good to use `None`, but `Ob0` already uses `None` as its default value and the default in `Parameter` needs to be different. And it is not good to use `nan` because we'd like to be able to recognize the `Parameter` default by type, not just by value.

Now that the big picture is cleared up I just have handful of small comments.

File: astropy/cosmology/tests/test_parameter.py (hunk 4 above)
Comment (nstarman) [reply astropy/cosmology/tests/test_parameter.py:67]: Good idea.

File: astropy/cosmology/tests/test_parameter.py (hunk 5 above)
Comment (nstarman) [reply astropy/cosmology/tests/test_parameter.py:169]: Done!

File: astropy/cosmology/parameter/_core.py (hunk 3 above)
Comment (nstarman) [reply astropy/cosmology/parameter/_core.py:318]: Done!

File: docs/whatsnew/6.0.rst (hunk 6 above)
Comment (nstarman) [reply docs/whatsnew/6.0.rst:143]: I don't want to highlight `<MISSING>` over the benefits from `dataclass`

File: astropy/cosmology/tests/test_parameter.py (hunk 4 above)
Comment (nstarman) [reply astropy/cosmology/tests/test_parameter.py]: Done!

File: astropy/cosmology/parameter/_core.py [hunk 7]
```diff
@@ -70,6 +83,9 @@ class Parameter:
 
     Parameters
     ----------
+    default : Any (optional, keyword-only)
+        Default value of the Parameter. If ``MISSING`` (default), the
```
Comment (mhvk) [inline astropy/cosmology/parameter/_core.py:87]: I would write "If not given, the" - no need to mention the sentinel value here.

File: astropy/cosmology/parameter/_core.py [hunk 8]
```diff
@@ -100,6 +116,13 @@ class Parameter:
     if not PYTHON_LT_3_10:
         _: KW_ONLY
 
+    default: Any = MISSING
+    """Default value of the Parameter.
+
+    If ``MISSING`` (default), the Parameter must be set when initializing the
```
Comment (mhvk) [inline astropy/cosmology/parameter/_core.py:122]: Maybe
"""
By default, set to ``MISSING``, which indicates the parameter must be set
when initializing the cosmology.
"""

File: astropy/cosmology/tests/test_parameter.py [hunk 9]
```diff
@@ -139,15 +154,21 @@ def test_Parameter_derived(self, cosmo_cls, all_parameter):
             all_parameter.name not in cosmo_cls.__parameters__
         )
 
+    def test_Parameter_default(self, cosmo_cls, all_parameter):
+        """Test :attr:`astropy.cosmology.Parameter.default`."""
+        assert hasattr(all_parameter, "default")
+        assert isinstance(
```
Comment (mhvk) [inline astropy/cosmology/tests/test_parameter.py:160]: I think I'd write this as
```
assert all_parameter.default is MISSING or all_parameter.default is None or isinstance(all_parameter.default, (float, u.Quantity))
```
This tests something closer to what is actually done (and `Sentinel` is an implementation detail).

File: docs/whatsnew/6.0.rst (hunk 6 above)
Comment (mhvk) [reply docs/whatsnew/6.0.rst:143]: I'm with @eerovaher - just write it out. The `...` is more likely to lead to confusion.

Or just pick a parameter which does not have `MISSING` as its default.

**Review by mhvk — APPROVED** [review]
This looks great. My comments are really all nitpicks, about comments and docstrings. Approving since it is OK as is too.

File: docs/whatsnew/6.0.rst (hunk 6 above)
Comment (nstarman) [reply docs/whatsnew/6.0.rst:143]: I'll do the latter!

File: astropy/cosmology/tests/test_parameter.py (hunk 9 above)
Comment (nstarman) [reply astropy/cosmology/tests/test_parameter.py:160]: I like separating out the MISSING, but NoneType is py3.10+, so there's no need to separate that out. Ruff will auto-upgrade `type(None)` -> `NoneType` when we next bump versions.

**Comment by nstarman** [comment]
Thanks @eerovaher and @mhvk for the detailed back and forth! Let's squash this in 🚀

File: astropy/cosmology/parameter/_core.py (hunk 7 above)
Comment (nstarman) [reply astropy/cosmology/parameter/_core.py]: Done.

File: astropy/cosmology/parameter/_core.py (hunk 8 above)
Comment (nstarman) [reply astropy/cosmology/parameter/_core.py]: Done!

File: docs/whatsnew/6.0.rst (hunk 6 above)
Comment (nstarman) [reply docs/whatsnew/6.0.rst]: Done!

**Comment by nstarman** [comment]
Ping @eerovaher if you approve as well.

File: astropy/cosmology/parameter/_core.py [hunk 10]
```diff
@@ -275,8 +308,10 @@ def clone(self, **kw):
     def __repr__(self) -> str:
         """Return repr(self)."""
         fields_repr = (
+            # Get the repr, using the input fvalidate over the processed value
             f"{f.name}={(getattr(self, f.name if f.name != 'fvalidate' else '_fvalidate_in'))!r}"
             for f in fields(self)
-            if f.repr
+            # Only show fields that should be displayed and are not sentinel values
+            if (f.repr and (f.name != "default" or self.default is not MISSING))
```
Comment (eerovaher) [inline astropy/cosmology/parameter/_core.py:315]: The outer parenthesis should be removed, but once that's done the pull request can be merged.
```suggestion
            if f.repr and (f.name != "default" or self.default is not MISSING)
```

Comment (nstarman) [reply astropy/cosmology/parameter/_core.py]: 👍
